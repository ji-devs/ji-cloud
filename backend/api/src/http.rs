use std::{net::TcpListener, sync::Arc};

use actix_service::Service;
use actix_web::{
    dev::{MessageBody, Server, ServiceRequest, ServiceResponse},
    error::Error,
    middleware::{Compat, Condition},
    web::{method, Data},
    HttpResponse,
};
use core::{
    config::JSON_BODY_LIMIT,
    env::env_bool,
    http::{get_addr, get_tcp_fd},
    settings::RuntimeSettings,
};
use futures::Future;
use sqlx::postgres::PgPool;
use tracing::Span;
use tracing_actix_web::{root_span, DefaultRootSpanBuilder, RootSpanBuilder, TracingLogger};

use crate::{
    error::BasicError,
    service::{self, mail, s3, upload::cleaner, ServiceData},
    translate,
};

mod cors;
mod endpoints;

fn log_ise<B: MessageBody, T>(
    request: ServiceRequest,
    srv: &T,
) -> impl Future<Output = actix_web::Result<T::Response>>
where
    T: Service<ServiceRequest, Response = ServiceResponse<B>, Error = actix_web::Error>,
{
    let uri: serde_json::Value = request.uri().to_string().into();
    let method: serde_json::Value = request.method().to_string().into();

    let future = srv.call(request);
    async {
        let mut result = future.await?;
        if result.status() == 500 {
            let response: &mut actix_web::HttpResponse<_> = result.response_mut();

            if let Some(err) = response.extensions_mut().remove::<anyhow::Error>() {
                log::error!("ISE while responding to request: {:?}", err);
                sentry::add_breadcrumb(sentry::Breadcrumb {
                    ty: "http".to_owned(),
                    category: Some("request".into()),
                    data: {
                        let mut map = sentry::protocol::Map::new();
                        map.insert("url".to_owned(), uri);
                        map.insert("method".to_owned(), method);
                        map
                    },
                    ..Default::default()
                });

                sentry::integrations::anyhow::capture_anyhow(&err);
            }
        }

        Ok(result)
    }
}

pub struct Application {
    port: u16,
    server: Option<Server>,
}

impl Application {
    pub fn new(port: u16, server: Server) -> Self {
        Self {
            port,
            server: Some(server),
        }
    }

    pub fn port(&self) -> u16 {
        self.port
    }

    // A more expressive name that makes it clear that
    // this function only returns when the application is stopped.
    pub async fn run_until_stopped(mut self) -> Result<(), std::io::Error> {
        if let Some(server) = self.server.take() {
            server.await?;
        }

        Ok(())
    }

    pub async fn stop(mut self, graceful: bool) {
        if let Some(server) = self.server.take() {
            server.stop(graceful).await
        }
    }
}

impl Drop for Application {
    fn drop(&mut self) {
        if let Some(server) = self.server.take() {
            let _ = tokio::spawn(server.stop(false));
        }
    }
}

#[actix_web::main]
pub async fn build_and_run(
    pool: PgPool,
    settings: RuntimeSettings,
    s3: Option<s3::Client>,
    gcp_key_store: Option<service::GcpAccessKeyStore>,
    gcs: Option<service::storage::Client>,
    algolia: Option<crate::algolia::Client>,
    algolia_key_store: Option<crate::algolia::SearchKeyStore>,
    jwk_verifier: Arc<crate::jwk::JwkVerifier>,
    mail_client: Option<mail::Client>,
    algolia_manager: Option<crate::algolia::Manager>,
    media_upload_cleaner: Option<cleaner::UploadCleaner>,
    google_translate: Option<translate::GoogleTranslate>,
) -> anyhow::Result<()> {
    let app = build(
        pool,
        settings,
        s3,
        gcp_key_store,
        gcs,
        algolia,
        algolia_key_store,
        jwk_verifier,
        mail_client,
        algolia_manager,
        media_upload_cleaner,
        google_translate,
    )?;
    app.run_until_stopped().await?;

    Ok(())
}

pub fn build(
    pool: PgPool,
    settings: RuntimeSettings,
    s3: Option<s3::Client>,
    gcp_key_store: Option<service::GcpAccessKeyStore>,
    gcs: Option<service::storage::Client>,
    algolia: Option<crate::algolia::Client>,
    algolia_key_store: Option<crate::algolia::SearchKeyStore>,
    jwk_verifier: Arc<crate::jwk::JwkVerifier>,
    mail_client: Option<mail::Client>,
    algolia_manager: Option<crate::algolia::Manager>,
    media_upload_cleaner: Option<cleaner::UploadCleaner>,
    google_translate: Option<translate::GoogleTranslate>,
) -> anyhow::Result<Application> {
    let local_insecure = settings.is_local();
    let api_port = settings.api_port;

    let s3 = s3.map(ServiceData::new);
    let gcp_key_store = gcp_key_store.map(ServiceData::new);
    let gcs = gcs.map(ServiceData::new);
    let algolia = algolia.map(ServiceData::new);
    let algolia_key_store = algolia_key_store.map(ServiceData::new);
    let mail_client = mail_client.map(ServiceData::new);
    let algolia_manager = algolia_manager.map(ServiceData::new);
    let media_upload_cleaner = media_upload_cleaner.map(ServiceData::new);
    let google_translate = google_translate.map(ServiceData::new);

    let server = actix_web::HttpServer::new(move || {
        let app = actix_web::App::new()
            .app_data(Data::new(pool.clone()))
            .app_data(Data::new(settings.clone()));

        let app = match s3.clone() {
            Some(s3) => app.app_data(s3),
            None => app,
        };

        let app = match gcp_key_store.clone() {
            Some(gcp_key_store) => app.app_data(gcp_key_store),
            None => app,
        };

        let app = match gcs.clone() {
            Some(gcs) => app.app_data(gcs),
            None => app,
        };

        let app = match algolia.clone() {
            Some(algolia) => app.app_data(algolia),
            None => app,
        };

        let app = match algolia_key_store.clone() {
            Some(algolia_key_store) => app.app_data(algolia_key_store),
            None => app,
        };

        let app = match mail_client.clone() {
            Some(mail_client) => app.app_data(mail_client),
            None => app,
        };

        let app = match algolia_manager.clone() {
            Some(algolia_manager) => app.app_data(algolia_manager),
            None => app,
        };

        let app = match media_upload_cleaner.clone() {
            Some(media_upload_cleaner) => app.app_data(media_upload_cleaner),
            None => app,
        };
        let app = match google_translate.clone() {
            Some(google_translate) => app.app_data(google_translate),
            None => app,
        };

        let enable_tracing_logs = env_bool("ENABLE_TRACING_LOGS");

        app.app_data(Data::from(jwk_verifier.clone()))
            .wrap(cors::get(local_insecure))
            .wrap(Condition::new(
                !enable_tracing_logs,
                Compat::new(actix_web::middleware::Logger::default()),
            ))
            .wrap(Condition::new(
                enable_tracing_logs,
                Compat::new(TracingLogger::<JigziSpanBuilder>::new()),
            ))
            .wrap_fn(log_ise)
            .app_data(
                actix_web::web::JsonConfig::default()
                    .limit(JSON_BODY_LIMIT as usize)
                    .error_handler(|_, _| bad_request_handler()),
            )
            .app_data(
                actix_web::web::QueryConfig::default().error_handler(|_, _| bad_request_handler()),
            )
            .app_data(
                actix_web::web::PathConfig::default().error_handler(|_, _| bad_request_handler()),
            )
            .external_resource(
                "google_cloud_oauth",
                "https://accounts.google.com/o/oauth2/v2/auth",
            )
            .default_service(actix_web::web::to(default_route))
            .configure(endpoints::user::configure)
            .configure(endpoints::category::configure)
            .configure(endpoints::image::configure)
            .configure(endpoints::audio::configure)
            .configure(endpoints::meta::configure)
            .configure(endpoints::jig::configure)
            .configure(endpoints::module::configure)
            .configure(endpoints::additional_resource::configure)
            .configure(endpoints::jig::report::configure)
            .configure(endpoints::jig::curation::configure)
            .configure(endpoints::course::configure)
            .configure(endpoints::admin::configure)
            .configure(endpoints::animation::configure)
            .configure(endpoints::search::configure)
            .configure(endpoints::media::configure)
            .configure(endpoints::session::configure)
            .configure(endpoints::locale::configure)
            .configure(endpoints::scheduler::configure)
            .configure(endpoints::pdf::configure)
            .configure(endpoints::badge::configure)
            .route("/", method(http::Method::GET).to(no_content_response))
    });

    // if listenfd doesn't take a TcpListener (i.e. we're not running via
    // the command above), we fall back to explicitly binding to a given
    // host:port.
    let listener = if let Some(l) = get_tcp_fd() {
        l
    } else {
        TcpListener::bind(get_addr(Some(api_port)))?
    };

    let port = listener.local_addr().unwrap().port();

    let server = server.listen(listener)?;

    Ok(Application::new(port, server.run()))
}

fn default_route() -> HttpResponse {
    HttpResponse::NotFound().json(BasicError::with_message(
        http::StatusCode::NOT_FOUND,
        "Route not found".to_owned(),
    ))
}

async fn no_content_response() -> HttpResponse {
    HttpResponse::NoContent().finish()
}

pub fn bad_request_handler() -> actix_web::Error {
    BasicError::new(http::StatusCode::BAD_REQUEST).into()
}

pub struct JigziSpanBuilder;

impl RootSpanBuilder for JigziSpanBuilder {
    fn on_request_start(request: &ServiceRequest) -> Span {
        let route = request.match_pattern().unwrap_or("default".to_string());
        let transaction_name = format!("{} {}", request.method(), route);
        let transaction_name = transaction_name.as_str();
        root_span!(request, message = transaction_name)
    }

    fn on_request_end<B>(span: Span, outcome: &Result<ServiceResponse<B>, Error>) {
        DefaultRootSpanBuilder::on_request_end(span, outcome)
    }
}
