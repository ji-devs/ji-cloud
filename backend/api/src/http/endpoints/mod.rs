pub mod additional_resource;
pub mod admin;
pub mod animation;
pub mod audio;
pub mod category;
pub mod image;
pub mod jig;
pub mod locale;
pub mod media;
pub mod meta;
pub mod module;
pub mod search;
pub mod session;
pub mod user;
