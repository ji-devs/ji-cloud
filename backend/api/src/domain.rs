use actix_web::{
    cookie::{Cookie, SameSite},
    HttpResponse, Responder,
};
use chrono::{DateTime, Utc};
use http::StatusCode;
use shared::domain::user::UserScope;
use shared::domain::{
    category::{Category, CategoryId},
    session::AUTH_COOKIE_NAME,
};
use std::convert::TryFrom;
use std::{collections::HashMap, fmt};
use uuid::Uuid;

#[derive(Debug)]
pub struct RawCategory {
    pub id: Uuid,
    pub parent_id: Option<Uuid>,
    pub name: String,
    pub index: i16,
    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>,
    pub image_count: i64,
    pub jig_count: i64,
    pub user_scopes: Vec<i16>,
}

pub fn build_tree(mut categories: Vec<RawCategory>) -> Vec<Category> {
    let mut parent_to_id_index: HashMap<Option<Uuid>, Vec<usize>> = HashMap::new();

    for (idx, category) in categories.iter().enumerate() {
        parent_to_id_index
            .entry(category.parent_id)
            .and_modify(|it| it.push(idx))
            .or_insert_with(|| vec![idx]);
    }

    parent_to_id_index
        .values_mut()
        .for_each(|it| it.sort_unstable_by_key(|v| categories[*v].index));

    build_tree_recursive(&parent_to_id_index, &mut categories, None)
}

fn build_tree_recursive(
    parent_to_id_index: &HashMap<Option<Uuid>, Vec<usize>>,
    categories: &mut Vec<RawCategory>,
    seed_id: Option<Uuid>,
) -> Vec<Category> {
    let indices = match parent_to_id_index.get(&seed_id) {
        Some(indecies) => indecies,
        None => return Vec::new(),
    };

    indices
        .iter()
        .copied()
        .map(|category_index| {
            let children = build_tree_recursive(
                parent_to_id_index,
                categories,
                Some(categories[category_index].id),
            );

            let raw: &mut RawCategory = &mut categories[category_index];

            Category {
                id: CategoryId(raw.id),
                name: raw.name.clone(),
                created_at: raw.created_at,
                updated_at: raw.updated_at,
                image_count: raw.image_count as u64,
                jig_count: raw.jig_count as u64,
                user_scopes: {
                    raw.user_scopes
                        .clone()
                        .into_iter()
                        .map(|x| UserScope::try_from(x).expect("detected an invalid user scope"))
                        .collect()
                },
                children,
            }
        })
        .collect()
}

#[derive(Debug, Copy, Clone, sqlx::Type)]
#[repr(u16)]
pub enum RegistrationStatus {
    /// The user was just registered ([`POST /v1/user`](shared::api::endpoints::user::Register))
    ///
    /// This state is *skipped* via ouath (they're automatically [`Validated`](Self::Validated))
    /// In this state the user is allowed to:
    /// * Request a new verification email
    /// * Request a password reset(?)
    New = 0,

    /// The user has gone through OAuth and created a new user, or has finished email verification.
    ///
    /// In this state the user is allowed to:
    /// * Request a password reset
    /// * Complete registration via creating their profile
    /// * Delete their account
    Validated = 1,

    /// The user has *completely* finished signup, and their profile has been completed.
    ///
    /// In this state the user is allowed to do anything that their scopes allow.
    /// All users can:
    /// * Change their email (not implemented nor currently on the roadmap)
    /// * Change their password if logged in, reset their password if not.
    /// * Access the API
    /// * Change their profile details (not currently implemented)
    /// * Delete their account
    Complete = 2,
}

#[derive(Debug)]
pub struct NoContentClearAuth;

impl fmt::Display for NoContentClearAuth {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("No Content")
    }
}

impl Responder for NoContentClearAuth {
    fn respond_to(self, _: &actix_web::HttpRequest) -> HttpResponse {
        let mut cookie = Cookie::named(AUTH_COOKIE_NAME);
        cookie.set_max_age(time::Duration::seconds(0));
        cookie.set_http_only(true);
        cookie.set_same_site(SameSite::Strict);

        actix_web::HttpResponse::build(StatusCode::NO_CONTENT)
            .content_type("application/octet-stream")
            .cookie(cookie)
            .finish()
    }
}
