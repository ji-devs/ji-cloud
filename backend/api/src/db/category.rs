use crate::{
    domain::{build_tree, RawCategory},
    error::{self, Delete},
};
use futures::TryStreamExt;
use shared::domain::category::{Category, CategoryId};
use shared::domain::user::UserScope;
use sqlx::{Executor, PgPool};
use std::convert::TryFrom;
use tracing::instrument;
use uuid::Uuid;

#[instrument(skip_all)]
pub async fn get_top_level(db: &sqlx::PgPool) -> anyhow::Result<Vec<Category>> {
    sqlx::query!(
        r#"
select id                                                                 as "id: CategoryId",
       name,
       created_at,
       updated_at,
       (select count(*)::int8 from image_category where category_id = id) as "image_count!",
       (select count(*)::int8 from jig_data_category where category_id = id) as "jig_count!",
       user_scopes
from category
where parent_id is null
order by index
 "#
    )
    .fetch(db)
    .map_ok(|it| Category {
        id: it.id,
        created_at: it.created_at,
        updated_at: it.updated_at,
        name: it.name,
        children: vec![],
        image_count: it.image_count as u64,
        jig_count: it.jig_count as u64,
        user_scopes: {
            let scopes = it
                .user_scopes
                .into_iter()
                .map(|x| UserScope::try_from(x).expect("failed to convert to UserScope"))
                .collect();

            scopes
        },
    })
    .try_collect()
    .await
    .map_err(Into::into)
}

#[instrument(skip(db))]
pub async fn get_exact(db: &sqlx::PgPool, ids: &[Uuid]) -> sqlx::Result<Vec<Category>> {
    sqlx::query!(
        //language=SQL
        r#"
select id                                                                 as "id!: CategoryId",
       name                                                               as "name!",
       created_at                                                         as "created_at!",
       updated_at,
       (select count(*)::int8 from image_category where category_id = id) as "image_count!",
       (select count(*)::int8 from jig_data_category where category_id = id)   as "jig_count!",
       user_scopes                                                        as "user_scopes!"
from category
         inner join unnest($1::uuid[]) with ordinality t(id, ord) USING (id)
order by t.ord
"#,
        ids
    )
    .fetch(db)
    .map_ok(|it| Category {
        id: it.id,
        name: it.name,
        created_at: it.created_at,
        updated_at: it.updated_at,
        children: vec![],
        image_count: it.image_count as u64,
        jig_count: it.jig_count as u64,
        user_scopes: {
            it.user_scopes
                .into_iter()
                .map(|x| UserScope::try_from(x).expect("failed to convert to UserScope"))
                .collect::<Vec<_>>()
        },
    })
    .try_collect()
    .await
}

#[instrument(skip(db))]
pub async fn get_subtree(db: &sqlx::PgPool, ids: &[Uuid]) -> sqlx::Result<Vec<Category>> {
    sqlx::query_file_as!(RawCategory, "query/category/get_subtree.sql", ids)
        .fetch_all(db)
        .await
        .map(build_tree)
}

#[instrument(skip_all)]
pub async fn get_tree(db: &sqlx::PgPool) -> sqlx::Result<Vec<Category>> {
    sqlx::query_as!(
        RawCategory,
        r#"
select id,
       parent_id,
       name,
       index,
       created_at,
       updated_at,
       (select count(*) from image_category where category_id = id)::int8 as "image_count!",
       (select count(*) from jig_data_category where category_id = id)::int8 as "jig_count!",
       user_scopes
from category
"#
    )
    .fetch_all(db)
    .await
    .map(build_tree)
}

#[instrument(skip_all)]
pub async fn get_ancestor_tree(db: &sqlx::PgPool, ids: &[Uuid]) -> sqlx::Result<Vec<Category>> {
    sqlx::query_file_as!(RawCategory, "query/category/get_ancestor_tree.sql", ids)
        .fetch_all(db)
        .await
        .map(build_tree)
}

#[instrument(skip(db))]
pub async fn create(
    db: &sqlx::PgPool,
    name: &str,
    parent_id: Option<CategoryId>,
) -> sqlx::Result<(CategoryId, u16)> {
    let res = sqlx::query!(
        r#"
insert into category (index, parent_id, name, user_scopes)
VALUES((select count(*)::int2 from category where parent_id is not distinct from $1), $1, $2, array[]::smallint[])
returning index, id"#,
        parent_id.map(|it| it.0),
        name,
    )
    .fetch_one(db)
    .await?;

    Ok((CategoryId(res.id), res.index as u16))
}

/// checks if moving the category with `id` to have `new_parent` as it's parent would create a cycle
///
/// A cycle is where there's a category that is it's own descendant.
async fn would_cycle(
    txn: &mut sqlx::PgConnection,
    id: Uuid,
    new_parent: Uuid,
) -> sqlx::Result<bool> {
    // trivially having `self` as your parent would cycle, as you would have a child that's yourself.
    if new_parent == id {
        return Ok(true);
    }

    sqlx::query!(
        r#"
with recursive cte(parent_id) as (
select parent_id from category where id = $1
union all
select c.parent_id from category c inner join cte on cte.parent_id = c.id
) select exists(select 1 from cte where parent_id = $2) as "would_cycle!"
    "#,
        new_parent,
        id
    )
    .fetch_one(txn)
    .await
    .map(|res| res.would_cycle)
}

enum UpdateLoopError {
    UpdateError(error::CategoryUpdate),
    Sqlx(sqlx::Error),
}

impl From<sqlx::Error> for UpdateLoopError {
    fn from(err: sqlx::Error) -> Self {
        Self::Sqlx(err)
    }
}

impl From<error::CategoryUpdate> for UpdateLoopError {
    fn from(err: error::CategoryUpdate) -> Self {
        Self::UpdateError(err)
    }
}

async fn update_slow(
    db: &sqlx::PgPool,
    id: Uuid,
    new_parent: Option<Option<Uuid>>,
    name: Option<&str>,
    index: Option<i16>,
) -> Result<(), UpdateLoopError> {
    let mut txn = db.begin().await?;

    txn.execute("set transaction isolation level repeatable read")
        .await?;

    let category_info = sqlx::query!(
        r#"
select parent_id, index from category where id = $1 for update
    "#,
        id
    )
    .fetch_optional(&mut txn)
    .await?
    .ok_or(error::CategoryUpdate::CategoryNotFound)?;

    if let Some(name) = name {
        #[allow(clippy::suspicious_else_formatting)]
        sqlx::query!("update category set name = $1 where id = $2", name, id)
            .execute(&mut txn)
            .await?;
    }

    let mut current_index = category_info.index;
    if let Some(parent_id) = new_parent {
        if parent_id != category_info.parent_id {
            // check that the new parent isn't a descendant (to avoid cycles)
            if let Some(new_parent) = parent_id {
                let would_cycle = would_cycle(&mut txn, id, new_parent).await?;

                if would_cycle {
                    return Err(error::CategoryUpdate::Cycle.into());
                }
            }

            // handle the new parent not existing and return `ParentCategoryNotFound`
            let res = sqlx::query!(
                r#"
update category
set parent_id = $1,
    updated_at = now(),
    index = (select count(*)::int2 from category where parent_id is not distinct from $1)
where id = $2
returning index
"#,
                parent_id,
                id
            )
            .fetch_optional(&mut txn)
            .await?
            .ok_or(error::CategoryUpdate::ParentCategoryNotFound)?;

            current_index = res.index;

            backshift(&mut txn, category_info.parent_id, category_info.index, None).await?;
        }
    }

    if let Some(new_index) = index {
        let current_parent = new_parent.unwrap_or(category_info.parent_id);
        if new_index < current_index {
            sqlx::query!(
                r#"
update category
set updated_at = now(),
    index = index + 1
where index >= $1 and index < $2 and parent_id is not distinct from $3
                "#,
                new_index,
                current_index,
                current_parent
            )
            .execute(&mut txn)
            .await?;
        }

        if new_index > current_index {
            backshift(&mut txn, current_parent, current_index, Some(new_index)).await?;
        }

        if new_index != current_index {
            sqlx::query!(
                r#"
update category
set updated_at = now(),
    index = least((select count(*)::int2 from category c where c.parent_id is not distinct from parent_id), $1)
where id = $2
"#,
                index,
                id
            )
            .execute(&mut txn)
            .await?;
        }
    }

    txn.commit().await?;

    Ok(())
}

#[instrument(skip(db))]
pub async fn update(
    db: &sqlx::PgPool,
    CategoryId(id): CategoryId,
    parent_id: Option<Option<CategoryId>>,
    name: Option<&str>,
    index: Option<i16>,
    user_scopes: Option<Vec<UserScope>>,
) -> Result<(), error::CategoryUpdate> {
    // fast track for if we're only updating the `name`:
    // the reasoning is due to an observation:
    // * we have to have retry logic for anything that involves transactions here (which updating the parent id and index _requires_)
    // * This is a single query, which can execute faster than opening a transaction and doing all the other stuff does.
    if parent_id.is_none() && index.is_none() {
        if let Some(name) = name {
            let rows_updated = sqlx::query!(
                "update category set name = $1, updated_at = now() where id = $2",
                name,
                id
            )
            .execute(db)
            .await?
            .rows_affected();

            match rows_updated {
                0 => return Err(error::CategoryUpdate::CategoryNotFound),
                1 => {}
                _ => unreachable!(),
            }
        }

        if let Some(user_scopes) = user_scopes {
            let scopes: Vec<i16> = user_scopes.iter().map(|x| *x as i16).collect();

            let rows_updated = sqlx::query!(
                "update category set user_scopes= $1, updated_at = now() where id = $2",
                scopes.as_slice(),
                id
            )
            .execute(db)
            .await?
            .rows_affected();

            match rows_updated {
                0 => return Err(error::CategoryUpdate::CategoryNotFound),
                1 => {}
                _ => unreachable!(),
            }
        }

        // Regardless of if the name is updated,
        // we know that neither parent_id nor index are going to be updated,
        // so we can just return here.
        return Ok(());
    }

    // note: at this point, at least one of `parent_id` or `index` are not null,
    // however, we don't know if we will end up updating either (they could be the same)

    let new_parent = parent_id.map(|id| id.map(|it| it.0));
    loop {
        match update_slow(db, id, new_parent, name, index).await {
            Ok(()) => return Ok(()),
            Err(UpdateLoopError::UpdateError(e)) => return Err(e),
            Err(UpdateLoopError::Sqlx(sqlx::Error::Database(e)))
                if e.code().as_deref() == Some("40001") => {}
            Err(UpdateLoopError::Sqlx(e)) => {
                return Err(error::CategoryUpdate::InternalServerError(e.into()))
            }
        }
    }
}

async fn backshift(
    txn: &mut sqlx::PgConnection,
    parent_id: Option<Uuid>,
    old_index: i16,
    new_index: Option<i16>,
) -> sqlx::Result<()> {
    sqlx::query!(
        r#"
update category
set index = index - 1, updated_at = now()
where index > $1 and index <= $2 is not false and parent_id is not distinct from $3
"#,
        old_index,
        new_index,
        parent_id
    )
    .execute(txn)
    .await
    .map(drop)
}

pub async fn delete(db: &PgPool, id: CategoryId) -> Result<(), Delete> {
    async fn inner(db: &PgPool, id: CategoryId) -> sqlx::Result<()> {
        let mut txn = db.begin().await?;

        txn.execute("set transaction isolation level repeatable read")
            .await?;

        let res = sqlx::query!(
            "delete from category where id = $1 returning index, parent_id",
            id.0
        )
        .fetch_optional(&mut txn)
        .await?;

        if let Some(res) = res {
            backshift(&mut txn, res.parent_id, res.index, None).await?;
        }

        txn.commit().await?;

        Ok(())
    }

    loop {
        match inner(db, id).await {
            Ok(()) => return Ok(()),
            Err(sqlx::Error::Database(e)) if e.code().as_deref() == Some("40001") => {}
            Err(e) => return Err(e.into()),
        }
    }
}
