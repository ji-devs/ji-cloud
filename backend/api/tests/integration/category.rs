use http::StatusCode;
use serde_json::json;
use shared::domain::category::{
    CategoryTreeScope, CreateCategoryRequest, GetCategoryRequest, NewCategoryResponse,
};
use uuid::Uuid;

use crate::{
    fixture::Fixture,
    helpers::{initialize_server, LoginExt},
};

#[actix_rt::test]
async fn create() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/category", port))
        .login()
        .json(&CreateCategoryRequest {
            name: "One".to_owned(),
            parent_id: None,
        })
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let _body: NewCategoryResponse = resp.json().await?;

    app.stop(false).await;

    Ok(())
}

#[actix_rt::test]
async fn get() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::CategoryOrdering], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/category", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body);

    Ok(())
}

async fn get_nested_categories(query: &GetCategoryRequest) -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::CategoryNesting], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/category", port))
        .query(query)
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body);

    Ok(())
}

#[actix_rt::test]
async fn nested_top_level() -> anyhow::Result<()> {
    get_nested_categories(&GetCategoryRequest::default()).await
}

#[actix_rt::test]
async fn nested_whole_tree() -> anyhow::Result<()> {
    get_nested_categories(&GetCategoryRequest {
        scope: Some(CategoryTreeScope::Decendants),
        ids: vec![],
    })
    .await
}

#[actix_rt::test]
async fn nested_overlapping() -> anyhow::Result<()> {
    get_nested_categories(&GetCategoryRequest {
        scope: Some(CategoryTreeScope::Decendants),
        ids: vec![
            "afbce03c-e90f-11ea-8281-cfde02f6b582".parse()?,
            "e315d3b2-e90f-11ea-8281-73cd69c14821".parse()?,
        ],
    })
    .await
}

#[actix_rt::test]
async fn nested_ancestors() -> anyhow::Result<()> {
    get_nested_categories(&GetCategoryRequest {
        scope: Some(CategoryTreeScope::Ancestors),
        ids: vec!["e315d3b2-e90f-11ea-8281-73cd69c14821".parse()?],
    })
    .await
}

#[actix_rt::test]
async fn nested_exact() -> anyhow::Result<()> {
    get_nested_categories(&GetCategoryRequest {
        scope: None,
        ids: vec![
            "afbce03c-e90f-11ea-8281-cfde02f6b582".parse()?,
            "01cff7d8-e910-11ea-8281-7f86c625a156".parse()?,
        ],
    })
    .await
}

#[actix_rt::test]
async fn upgdate_ordering() -> anyhow::Result<()> {
    let category_three = "81c4796a-e883-11ea-93f0-df2484ab6b11";

    let app = initialize_server(&[Fixture::User, Fixture::CategoryOrdering], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/category/{}",
            port, category_three
        ))
        .json(&json!({"index": 0}))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/category", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, {".**.updated_at" => "[timestamp]"});

    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/category/{}",
            port, category_three
        ))
        .json(&json!({"index": 2}))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/category", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body, {".**.updated_at" => "[timestamp]"});

    Ok(())
}

#[actix_rt::test]
async fn delete() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::CategoryOrdering], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .delete(&format!(
            "http://0.0.0.0:{}/v1/category/7fe19326-e883-11ea-93f0-5343493c17c4",
            port,
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/category", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body, {".**.updated_at" => "[timestamp]"});

    Ok(())
}

async fn update(id: Uuid, body: &serde_json::Value) -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::CategoryOrdering], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .patch(&format!("http://0.0.0.0:{}/v1/category/{}", port, id))
        .json(&body)
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/category?scope=Decendants",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body, {".**.updated_at" => "[timestamp]"});

    Ok(())
}

#[actix_rt::test]
async fn update_parent() -> anyhow::Result<()> {
    update(
        "7fe19326-e883-11ea-93f0-5343493c17c4".parse()?,
        &json!({"parent_id": "81c4796a-e883-11ea-93f0-df2484ab6b11"}),
    )
    .await
}

#[actix_rt::test]
async fn update_reparent_move() -> anyhow::Result<()> {
    update(
        "7fe19326-e883-11ea-93f0-5343493c17c4".parse()?,
        &json!({"parent_id": (), "index": 0}),
    )
    .await
}

#[actix_rt::test]
async fn update_move() -> anyhow::Result<()> {
    update(
        "81c4796a-e883-11ea-93f0-df2484ab6b11".parse()?,
        &json!({"index": 1}),
    )
    .await
}

#[actix_rt::test]
async fn update_rename() -> anyhow::Result<()> {
    update(
        "81c4796a-e883-11ea-93f0-df2484ab6b11".parse()?,
        &json!({"name": "abc123"}),
    )
    .await
}
