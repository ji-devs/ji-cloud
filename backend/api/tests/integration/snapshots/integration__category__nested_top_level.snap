---
source: tests/integration/category.rs
expression: body

---
{
  "categories": [
    {
      "id": "afbce03c-e90f-11ea-8281-cfde02f6b582",
      "user_scopes": [
        "Admin",
        "ManageAnimation"
      ],
      "name": "Parent",
      "created_at": "2020-08-28T09:20:17.039205Z",
      "updated_at": null,
      "image_count": 0,
      "jig_count": 0
    }
  ]
}
